SeaPublicWebsite.UnitTests/PropertyDataServiceTests.cs

```diff
@@ -0,0 +1,108 @@
+            LoftAccess = LoftAccess.No,
+            RoofInsulated = RoofInsulated.No,
+            HasOutdoorSpace = HasOutdoorSpace.No,
+            GlazingType = GlazingType.Both,
+            HeatingType = HeatingType.Other,
+            OtherHeatingType = OtherHeatingType.Biomass,
+            HasHotWaterCylinder = HasHotWaterCylinder.No,
+            NumberOfOccupants = 2,
+            HeatingPattern = HeatingPattern.Other,
+            HoursOfHeatingMorning = 2,
+            HoursOfHeatingEvening = 2,
+            Temperature = 20,
+            UneditedData = new PropertyData(),
+            HasSeenRecommendations = false,
+            RecommendationsFirstRetrievedAt = dateTime,
+            PropertyRecommendations = null
+        };
+    }
+    
+    public PropertyDataServiceTests()
+    {
+        mockPropertyDataStore = new Mock<IPropertyDataStore>();
+        mockRecommendationService = new Mock<IRecommendationService>();
+        underTest = new PropertyDataService(mockPropertyDataStore.Object, mockRecommendationService.Object);
+    }
+
+    [Test]
+    public async Task UpdatePropertyDataWithRecommendations_WhenCalledOnReferenceThatAlreadyHasTimestamp_ReturnsPropertyDataWithSameTimestamp()
+    {
+        // Arrange
+        var testTime = new DateTime(2000, 1, 1);
+        mockRecommendationService.Setup(rs => rs.GetRecommendationsForPropertyAsync(It.IsAny<PropertyData>()))
+            .ReturnsAsync(new List<BreRecommendation>());
+        mockPropertyDataStore.Setup(ds => ds.LoadPropertyDataAsync("222222"))
+            .ReturnsAsync(InitializePropertyDataWithRecommendationsFirstRetrievedAt(testTime));
+        
+        // Act
+        var returnedPropertyData = await underTest.UpdatePropertyDataWithRecommendations("222222");
+        
+        // Assert
+        Assert.AreEqual(testTime, returnedPropertyData.RecommendationsFirstRetrievedAt);
+    }
+    
+    [Test]
+    public async Task UpdatePropertyDataWithRecommendations_WhenCalledOnReferenceThatHasNoTimestamp_ReturnsPropertyDataWithTimestampForNow()
+    {
+        // Arrange
+        var startTime = DateTime.Now.ToUniversalTime();
+        mockRecommendationService.Setup(rs => rs.GetRecommendationsForPropertyAsync(It.IsAny<PropertyData>()))
+            .ReturnsAsync(new List<BreRecommendation>());
+        mockPropertyDataStore.Setup(ds => ds.LoadPropertyDataAsync("222222"))
+            .ReturnsAsync(InitializePropertyDataWithRecommendationsFirstRetrievedAt(null));
+        mockPropertyDataStore.Setup(ds => ds.SavePropertyDataAsync(It.IsAny<PropertyData>()));
+        
+        // Act
+        var returnedPropertyData = await underTest.UpdatePropertyDataWithRecommendations("222222");
+
+        
+        // Assert
+        var endTime = DateTime.Now.ToUniversalTime();
+        Assert.LessOrEqual(startTime, returnedPropertyData.RecommendationsFirstRetrievedAt);
+        Assert.GreaterOrEqual(endTime, returnedPropertyData.RecommendationsFirstRetrievedAt);
```

Comment:
I'm not quite understanding the first bit, do you mean that when I mock this out, I deliver a custom time when DateTime.Now is called? Or capture DateTime.Now when it's called in the mock?

Then check that against what comes back from the method I'm testing?

Reply:
So you'd have something like
```
public class DateTimeProvider
{
  public DateTime Now()
  {
    return DateTime.Now();
  }
}
```
Then you inject your `DateTimeProvider` into your other class and can mock it out in your tests: set up so that `mockDateTimeProvider.Now` returns a datetime you specify, and then you can assert against that.

(this is just one option!)

Reply:
I get it now! :) Thanks for explaining